react-native 0.72.4 compatibility

**Which package/packages do you use?**

- [ ] `@stream-io/video-react-sdk`
- [x] `@stream-io/video-react-native-sdk`
- [ ] `@stream-io/video-client`

**Describe the bug**
Our RN app us using RN `0.7.2.4`, and after upgrading to `@stream-io/video-react-native-sdk` `1.9.1` the Android build is broken - a Java file fails to compile:
```
> Task :stream-io_video-react-native-sdk:compileDebugKotlin FAILED
'compileDebugJavaWithJavac' task (current target is 11) and 'compileDebugKotlin' task (current target is 17) jvm target compatibility should be set to the same Java version.
e: /Users/mccraigmccraig/dev/sona/sona/native-app/node_modules/@stream-io/video-react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt: (61, 30): Unresolved reference: reactDelegate
```

error is from this source: https://github.com/GetStream/stream-video-js/blob/main/packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt#L61

and the unresolved reference is here in `ReactActivity.java` from RN `main/HEAD`: https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactActivity.java#L68

but is missing in `ReactActivity` from RN `0.72.4`: https://github.com/facebook/react-native/blob/v0.72.4/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactActivity.java

so it seems `@stream-io/video-react-native-sdk:1.9.1` is not backwards compatible with RN `0.72.4`

**To Reproduce**
Steps to reproduce the behavior:

1. on a stream-video app on RN `0.72.4`
2. upgrade`@stream-io/video-react-native-sdk` to `1.9.1`
3. run Android build
4. BOOM!

**Expected behavior**
React Native `0.72.4` app using `@stream-io/video-react-native-sdk` `1.9.1` should build and run correctly on Android

**Screenshots**
n/a

**Version**
`@stream-io/video-react-native-sdk` `1.9.1`

- SDK version
1.9.1

**Env:**

**React Native**

- OS: Android
- Device or emulator: failed during compile

**Additional context**
n/a

Hello @oliverlaz I'm also getting the same issue with my app. 

Steps to reproduce the behavior:

1. Run on a RN 0.73.6
2. Upgrade @stream-io/video-react-native-sdk to 1.9.2 and @stream-io/react-native-webrtc to 125.0.3
3. Run on Android build but failed to compile

**FAILURE: Build failed with an exception.**

* What went wrong:
Execution failed for task ':stream-io_video-react-native-sdk:compileDebugKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

--- a/.github/workflows/react-native-workflow.yml
+++ b/.github/workflows/react-native-workflow.yml
@@ -86,7 +86,7 @@ jobs:
         run: bundle exec fastlane build_ios
 
       - name: Upload .ipa
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: ipa
           path: |
@@ -116,7 +116,7 @@ jobs:
         working-directory: sample-apps/react-native/dogfood
         run: bundle exec fastlane test_ios
 
-      - uses: actions/upload-artifact@v3
+      - uses: actions/upload-artifact@v4
         if: failure()
         with:
           name: iOS Test Data
@@ -184,13 +184,13 @@ jobs:
         run: bundle exec fastlane build_android_play_store
 
       - name: Upload .apk
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: apk
           path: '**/dist/*.apk'
 
       - name: Upload .aab
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           name: aab
           path: '**/dist/*.aab'
@@ -253,7 +253,7 @@ jobs:
           working-directory: sample-apps/react-native/dogfood
           script: bundle exec fastlane test_android
 
-      - uses: actions/upload-artifact@v3
+      - uses: actions/upload-artifact@v4
         if: failure()
         with:
           name: Android Test Data

--- a/packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt
+++ b/packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/StreamVideoReactNativeModule.kt
@@ -58,7 +58,7 @@ class StreamVideoReactNativeModule(reactContext: ReactApplicationContext) : Reac
                     }
                     // NOTE: workaround - on PiP mode, android goes to "paused but can render" state
                     // RN pauses rendering in paused mode, so we instruct it to resume here
-                    activity.reactDelegate?.onHostResume()
+                    reactApplicationContext?.onHostResume(activity)
                 }
             }
         }

--- a/packages/react-native-sdk/expo-config-plugin/__tests__/withAppDelegate.test.ts
+++ b/packages/react-native-sdk/expo-config-plugin/__tests__/withAppDelegate.test.ts
@@ -1,5 +1,5 @@
 import { getFixture } from '../fixtures/index';
-import withPushAppDelegate from '../src/withPushAppDelegate';
+import withAppDelegate from '../src/withAppDelegate';
 import { ExpoConfig } from '@expo/config-types';
 import { ConfigProps } from '../src/common/types';
 
@@ -41,10 +41,7 @@ describe('withStreamVideoReactNativeSDKAppDelegate', () => {
     };
 
     const props: ConfigProps = {};
-    const updatedConfig = withPushAppDelegate(
-      config,
-      props
-    ) as CustomExpoConfig;
+    const updatedConfig = withAppDelegate(config, props) as CustomExpoConfig;
 
     expect(
       updatedConfig.modResults.contents === config.modResults.contents
@@ -64,17 +61,21 @@ describe('withStreamVideoReactNativeSDKAppDelegate', () => {
     };
 
     const props: ConfigProps = {
+      iOSEnableMultitaskingCameraAccess: true,
       ringingPushNotifications: {
         disableVideoIos: true,
         includesCallsInRecentsIos: true,
       },
     };
 
-    const updatedConfig = withPushAppDelegate(
-      config,
-      props
-    ) as CustomExpoConfig;
+    const updatedConfig = withAppDelegate(config, props) as CustomExpoConfig;
 
+    expect(updatedConfig.modResults.contents).toMatch(
+      /#import <WebRTCModuleOptions.h>/
+    );
+    expect(updatedConfig.modResults.contents).toMatch(
+      /options.enableMultitaskingCameraAccess = YES/
+    );
     expect(updatedConfig.modResults.contents).toMatch(/#import "RNCallKeep.h"/);
     expect(updatedConfig.modResults.contents).toMatch(
       /#import <PushKit\/PushKit.h>/
@@ -99,13 +100,14 @@ describe('withStreamVideoReactNativeSDKAppDelegate', () => {
 
   it('should not modify config if already added', () => {
     const props: ConfigProps = {
+      iOSEnableMultitaskingCameraAccess: true,
       ringingPushNotifications: {
         disableVideoIos: true,
         includesCallsInRecentsIos: true,
       },
     };
 
-    const updatedConfig = withPushAppDelegate(
+    const updatedConfig = withAppDelegate(
       modifiedConfig!,
       props
     ) as CustomExpoConfig;
@@ -132,7 +134,7 @@ describe('withStreamVideoReactNativeSDKAppDelegate', () => {
         includesCallsInRecentsIos: false,
       },
     };
-    expect(() => withPushAppDelegate(config, props)).toThrow();
+    expect(() => withAppDelegate(config, props)).toThrow();
 
     config = {
       name: 'test-app',
@@ -143,6 +145,6 @@ describe('withStreamVideoReactNativeSDKAppDelegate', () => {
         contents: ExpoModulesAppDelegate,
       },
     };
-    expect(() => withPushAppDelegate(config, props)).toThrow();
+    expect(() => withAppDelegate(config, props)).toThrow();
   });
 });

--- a/packages/react-native-sdk/expo-config-plugin/src/common/types.ts
+++ b/packages/react-native-sdk/expo-config-plugin/src/common/types.ts
@@ -12,5 +12,6 @@ export type ConfigProps =
       androidKeepCallAlive?: boolean;
       enableScreenshare?: boolean;
       appleTeamId?: string;
+      iOSEnableMultitaskingCameraAccess?: boolean;
     }
   | undefined;

--- a/packages/react-native-sdk/expo-config-plugin/src/index.ts
+++ b/packages/react-native-sdk/expo-config-plugin/src/index.ts
@@ -3,7 +3,7 @@ import {
   createRunOncePlugin,
   withPlugins,
 } from '@expo/config-plugins';
-import withPushAppDelegate from './withPushAppDelegate';
+import withAppDelegate from './withAppDelegate';
 import withAndroidPermissions from './withAndroidPermissions';
 import withAndroidManifest from './withAndroidManifest';
 import withiOSInfoPlist from './withiOSInfoPlist';
@@ -22,7 +22,7 @@ const withStreamVideoReactNativeSDK: ConfigPlugin<ConfigProps> = (
 ) => {
   return withPlugins(config, [
     // ios
-    () => withPushAppDelegate(config, props),
+    () => withAppDelegate(config, props),
     () => withiOSInfoPlist(config, props),
     () => withIosScreenCapture(config, props),
     // android

--- a/packages/react-native-sdk/expo-config-plugin/src/withAppDelegate.ts
+++ b/packages/react-native-sdk/expo-config-plugin/src/withAppDelegate.ts
@@ -1,4 +1,7 @@
-import { ConfigPlugin, withAppDelegate } from '@expo/config-plugins';
+import {
+  ConfigPlugin,
+  withAppDelegate as withAppDelegateUtil,
+} from '@expo/config-plugins';
 import {
   addObjcImports,
   insertContentsInsideObjcFunctionBlock,
@@ -15,39 +18,45 @@ const DID_UPDATE_PUSH_CREDENTIALS =
 const DID_RECEIVE_INCOMING_PUSH =
   'pushRegistry:didReceiveIncomingPushWithPayload:forType:withCompletionHandler:';
 
-const withPushAppDelegate: ConfigPlugin<ConfigProps> = (
-  configuration,
-  props
-) => {
-  return withAppDelegate(configuration, (config) => {
-    if (!props?.ringingPushNotifications) {
-      // user doesnt want to use ringing push notifications, so quit early
+const withAppDelegate: ConfigPlugin<ConfigProps> = (configuration, props) => {
+  return withAppDelegateUtil(configuration, (config) => {
+    if (
+      !props?.ringingPushNotifications &&
+      !props?.iOSEnableMultitaskingCameraAccess
+    ) {
+      // quit early if no change is necessary
       return config;
     }
     if (['objc', 'objcpp'].includes(config.modResults.language)) {
       try {
         // all the imports that are needed
-        config.modResults.contents = addObjcImports(
+        if (props?.ringingPushNotifications) {
+          config.modResults.contents = addObjcImports(
+            config.modResults.contents,
+            [
+              '"RNCallKeep.h"',
+              '<PushKit/PushKit.h>',
+              '"RNVoipPushNotificationManager.h"',
+              '"StreamVideoReactNative.h"',
+            ]
+          );
+
+          config.modResults.contents = addDidFinishLaunchingWithOptionsRinging(
+            config.modResults.contents,
+            props.ringingPushNotifications
+          );
+
+          config.modResults.contents = addDidUpdatePushCredentials(
+            config.modResults.contents
+          );
+
+          config.modResults.contents = addDidReceiveIncomingPushCallback(
+            config.modResults.contents
+          );
+        }
+        addDidFinishLaunchingWithOptions(
           config.modResults.contents,
-          [
-            '"RNCallKeep.h"',
-            '<PushKit/PushKit.h>',
-            '"RNVoipPushNotificationManager.h"',
-            '"StreamVideoReactNative.h"',
-          ]
-        );
-
-        config.modResults.contents = addDidFinishLaunchingWithOptions(
-          config.modResults.contents,
-          props.ringingPushNotifications
-        );
-
-        config.modResults.contents = addDidUpdatePushCredentials(
-          config.modResults.contents
-        );
-
-        config.modResults.contents = addDidReceiveIncomingPushCallback(
-          config.modResults.contents
+          props.iOSEnableMultitaskingCameraAccess
         );
         return config;
       } catch (error: any) {
@@ -64,6 +73,26 @@ const withPushAppDelegate: ConfigPlugin<ConfigProps> = (
 };
 
 function addDidFinishLaunchingWithOptions(
+  contents: string,
+  iOSEnableMultitaskingCameraAccess: boolean | undefined
+) {
+  if (iOSEnableMultitaskingCameraAccess) {
+    addObjcImports(contents, ['<WebRTCModuleOptions.h>']);
+    const setupMethod = `WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
+  options.enableMultitaskingCameraAccess = YES;`;
+    if (!contents.includes('options.enableMultitaskingCameraAccess = YES')) {
+      contents = insertContentsInsideObjcFunctionBlock(
+        contents,
+        DID_FINISH_LAUNCHING_WITH_OPTIONS,
+        setupMethod,
+        { position: 'tailBeforeLastReturn' }
+      );
+    }
+  }
+  return contents;
+}
+
+function addDidFinishLaunchingWithOptionsRinging(
   contents: string,
   ringingPushNotifications: RingingPushNotifications
 ) {
@@ -183,4 +212,4 @@ function addDidReceiveIncomingPushCallback(contents: string) {
   return contents;
 }
 
-export default withPushAppDelegate;
+export default withAppDelegate;

--- a/packages/react-native-sdk/expo-config-plugin/src/withAppDelegate.ts
+++ b/packages/react-native-sdk/expo-config-plugin/src/withAppDelegate.ts
@@ -54,7 +54,7 @@ const withAppDelegate: ConfigPlugin<ConfigProps> = (configuration, props) => {
             config.modResults.contents
           );
         }
-        addDidFinishLaunchingWithOptions(
+        config.modResults.contents = addDidFinishLaunchingWithOptions(
           config.modResults.contents,
           props.iOSEnableMultitaskingCameraAccess
         );
@@ -77,9 +77,11 @@ function addDidFinishLaunchingWithOptions(
   iOSEnableMultitaskingCameraAccess: boolean | undefined
 ) {
   if (iOSEnableMultitaskingCameraAccess) {
-    addObjcImports(contents, ['<WebRTCModuleOptions.h>']);
+    contents = addObjcImports(contents, ['<WebRTCModuleOptions.h>']);
+
     const setupMethod = `WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
   options.enableMultitaskingCameraAccess = YES;`;
+
     if (!contents.includes('options.enableMultitaskingCameraAccess = YES')) {
       contents = insertContentsInsideObjcFunctionBlock(
         contents,

--- a/packages/react-native-sdk/ios/PictureInPicture/StreamPictureInPictureController.swift
+++ b/packages/react-native-sdk/ios/PictureInPicture/StreamPictureInPictureController.swift
@@ -151,7 +151,9 @@ import Foundation
     }
     
     @objc func cleanup() {
+        sourceView = nil
         contentViewController?.track = nil
+        contentViewController = nil
         if #available(iOS 15.0, *) {
             pictureInPictureController?.contentSource = nil
         }

--- a/packages/react-native-sdk/ios/RTCViewPip.swift
+++ b/packages/react-native-sdk/ios/RTCViewPip.swift
@@ -35,12 +35,13 @@ class RTCViewPip: UIView {
     }
     
     @objc func setStreamURL(_ streamReactTag: NSString) {
+        let tag = String(streamReactTag)
         webRtcModule?.workerQueue.async {
-            let stream = self.webRtcModule?.stream(forReactTag: String(streamReactTag))
+            let stream = self.webRtcModule?.stream(forReactTag: tag)
             let videoTracks = stream?.videoTracks ?? []
             let videoTrack = videoTracks.first
             if videoTrack == nil {
-                NSLog("PiP - No video stream for react tag: -\(streamReactTag)")
+                NSLog("PiP - No video stream for react tag: -\(tag)")
             } else {
                 DispatchQueue.main.async {
                     self.pictureInPictureController?.track = videoTrack

--- a/packages/react-native-sdk/src/components/Call/CallContent/RTCViewPipIOS.tsx
+++ b/packages/react-native-sdk/src/components/Call/CallContent/RTCViewPipIOS.tsx
@@ -5,7 +5,7 @@ import {
 } from '@stream-io/video-client';
 import { useCall, useCallStateHooks } from '@stream-io/video-react-bindings';
 import type { MediaStream } from '@stream-io/react-native-webrtc';
-import React, { useEffect } from 'react';
+import React, { useEffect, useMemo } from 'react';
 import {
   findNodeHandle,
   HostComponent,
@@ -22,44 +22,29 @@ type RTCViewPipNativeProps = {
   streamURL?: string;
 };
 
-// workaround to support hot reloading
-// https://medium.com/tribalscale/beyond-the-framework-using-react-native-with-swift-and-kotlin-cfccf4bb9a03
-let RTCViewPipNative: HostComponent<RTCViewPipNativeProps>;
-
-if (__DEV__) {
-  /* @ts-ignore */
-  const cachedView = global.RTCViewPipNative as
-    | HostComponent<RTCViewPipNativeProps>
-    | undefined;
-  if (!cachedView) {
-    RTCViewPipNative = requireNativeComponent(COMPONENT_NAME);
-    /* @ts-ignore */
-    global.RTCViewPipNative = RTCViewPipNative;
-  } else {
-    RTCViewPipNative = cachedView;
-  }
-} else {
-  RTCViewPipNative = requireNativeComponent(COMPONENT_NAME);
-}
+const RTCViewPipNative: HostComponent<RTCViewPipNativeProps> =
+  requireNativeComponent(COMPONENT_NAME);
 
 /** Wrapper for the native view
  * meant to stay private and not exposed */
-const RTCViewPip = React.forwardRef<
-  React.Ref<any>,
-  {
-    streamURL?: string;
-  }
->((props, ref) => {
-  if (Platform.OS !== 'ios') return null;
-  // @ts-ignore
-  return <RTCViewPipNative streamURL={props.streamURL} ref={ref} />;
-});
+const RTCViewPip = React.memo(
+  React.forwardRef<
+    React.Ref<any>,
+    {
+      streamURL?: string;
+    }
+  >((props, ref) => {
+    if (Platform.OS !== 'ios') return null;
+    // @ts-ignore
+    return <RTCViewPipNative streamURL={props.streamURL} ref={ref} />;
+  })
+);
 
 type Props = {
   includeLocalParticipantVideo?: boolean;
 };
 
-const RTCViewPipIOS = ({ includeLocalParticipantVideo }: Props) => {
+const RTCViewPipIOS = React.memo(({ includeLocalParticipantVideo }: Props) => {
   const call = useCall();
   const { useParticipants } = useCallStateHooks();
   const _allParticipants = useParticipants({
@@ -76,26 +61,7 @@ const RTCViewPipIOS = ({ includeLocalParticipantVideo }: Props) => {
       !includeLocalParticipantVideo;
   }, [includeLocalParticipantVideo]);
 
-  const [videoStreamToRender, setVideoStreamToRender] =
-    React.useState<MediaStream>();
-
   const nativeRef = React.useRef<any>(null);
-
-  React.useEffect(() => {
-    if (!participantInSpotlight) {
-      setVideoStreamToRender(undefined);
-      return;
-    }
-    const { videoStream, screenShareStream } = participantInSpotlight;
-
-    const isScreenSharing = hasScreenShare(participantInSpotlight);
-
-    const _videoStreamToRender = (isScreenSharing
-      ? screenShareStream
-      : videoStream) as unknown as MediaStream | undefined;
-
-    setVideoStreamToRender(_videoStreamToRender);
-  }, [participantInSpotlight]);
 
   React.useEffect(() => {
     let callClosedInvokedOnce = false;
@@ -130,9 +96,23 @@ const RTCViewPipIOS = ({ includeLocalParticipantVideo }: Props) => {
     };
   }, [call]);
 
-  return (
-    <RTCViewPip streamURL={videoStreamToRender?.toURL()} ref={nativeRef} />
-  );
-};
+  const streamURL = useMemo(() => {
+    if (!participantInSpotlight) {
+      return undefined;
+    }
+
+    const { videoStream, screenShareStream } = participantInSpotlight;
+
+    const isScreenSharing = hasScreenShare(participantInSpotlight);
+
+    const videoStreamToRender = (isScreenSharing
+      ? screenShareStream
+      : videoStream) as unknown as MediaStream | undefined;
+
+    return videoStreamToRender?.toURL();
+  }, [participantInSpotlight]);
+
+  return <RTCViewPip streamURL={streamURL} ref={nativeRef} />;
+});
 
 export default RTCViewPipIOS;

--- a/sample-apps/react-native/dogfood/ios/StreamReactNativeVideoSDKSample/AppDelegate.mm
+++ b/sample-apps/react-native/dogfood/ios/StreamReactNativeVideoSDKSample/AppDelegate.mm
@@ -125,6 +125,7 @@ continueUserActivity:(nonnull NSUserActivity *)userActivity
   center.delegate = self;
   
   WebRTCModuleOptions *options = [WebRTCModuleOptions sharedInstance];
+  options.enableMultitaskingCameraAccess = YES;
 //  uncomment below to see native webrtc logs
 //  options.loggingSeverity = RTCLoggingSeverityInfo;
   

--- a/sample-apps/react-native/dogfood/src/components/ActiveCall.tsx
+++ b/sample-apps/react-native/dogfood/src/components/ActiveCall.tsx
@@ -102,6 +102,7 @@ export const ActiveCall = ({
       />
       {!isInPiPMode && <CustomTopControls />}
       <CallContent
+        iOSPiPIncludeLocalParticipantVideo
         onHangupCallHandler={onHangupCallHandler}
         CallControls={CustomBottomControls}
         landscape={isLandscape}

